remotes/docker/authorizer.go

```diff
@@ -304,17 +304,22 @@ func (ah *authHandler) doBearerAuth(ctx context.Context) (token, refreshToken st
 				// As of September 2017, GCR is known to return 404.
 				// As of February 2018, JFrog Artifactory is known to return 401.
 				// As of January 2022, ACR is known to return 400.
-				if (errStatus.StatusCode == 405 && to.Username != "") || errStatus.StatusCode == 404 || errStatus.StatusCode == 401 || errStatus.StatusCode == 400 {
+				switch status := errStatus.StatusCode; {
+				case http.StatusMethodNotAllowed, http.StatusNotFound, http.StatusUnauthorized, http.StatusBadRequest:
+					if status == http.StatusMethodNotAllowed && to.Username == "" {
+						fallthrough
```

Comment:
I can have a look later to see if I can provide an alternative

Reply:
Ahh, can't be in a branch? I tested if it could fallthrough to default as I always forget, but didn't test the former.

Reply:
Also looking (will have to check git history) why we only check for the `to.Username` for this specific status code. Do we need to, or can we just continue, try authenticating, and make it fail there?

Reply:
I guess this could look like either;

```go
switch status := errStatus.StatusCode; status {
case http.StatusMethodNotAllowed, http.StatusNotFound, http.StatusUnauthorized, http.StatusBadRequest:
    if status != http.StatusMethodNotAllowed || (status == http.StatusMethodNotAllowed && to.Username != "") {
        resp, err := auth.FetchToken(ctx, ah.client, ah.header, to)
        if err != nil {
            return "", "", err
        }
        return resp.Token, resp.RefreshToken, nil
    }
}
log.G(ctx).WithFields(log.Fields{
    "status": errStatus.Status,
    "body":   string(errStatus.Body),
}).Debugf("token request failed")
```

Or of we prefer readability at the cost of some duplication:

```go
switch status := errStatus.StatusCode; status {
case http.StatusMethodNotAllowed:
    if to.Username != "" {
        resp, err := auth.FetchToken(ctx, ah.client, ah.header, to)
        if err != nil {
            return "", "", err
        }
        return resp.Token, resp.RefreshToken, nil
    }
case http.StatusNotFound, http.StatusUnauthorized, http.StatusBadRequest:
    resp, err := auth.FetchToken(ctx, ah.client, ah.header, to)
    if err != nil {
        return "", "", err
    }
    return resp.Token, resp.RefreshToken, nil
}
log.G(ctx).WithFields(log.Fields{
    "status": errStatus.Status,
    "body":   string(errStatus.Body),
}).Debugf("token request failed")
```

(Still need to look at the `to.Username` though)

Reply:
So, I did some initial digging;

So, it looks like;

- https://github.com/containerd/containerd/commit/9d3f452371b56c26036e822d848b83ef58e234b3 (https://github.com/containerd/containerd/pull/783) added a check for `405` **combined with a username** https://github.com/containerd/containerd/commit/9d3f452371b56c26036e822d848b83ef58e234b3#diff-d6dd575a9d10716db9ac385bf5f394a2e65d638311159e8726b3200fcc596989R444-R446
- https://github.com/containerd/containerd/commit/aaae7060672f3177c1d82c6119611e0a2cfbc023 (https://github.com/containerd/containerd/pull/1465) added a check for `404` because some registries return that status-code, but did _not_ add a check if a username was provided
- Now things probably because confusing, and when more status-codes were added, such as `401` in https://github.com/containerd/containerd/commit/373f1e56127de561c5254eac2656f9066ad24693 (https://github.com/containerd/containerd/pull/2171), that approach just got copied forward ("add a status code, but don't look at username").

From the above, I _suspect_ all of those status-codes _should_ be treated equal; either;

- if we receive one of those status-codes, then try to authenticate (irregardless if `username` was set or not)
- if we receive one of those status-codes, then _ONLY_ try to authenticate if a `username` was passed

That said, perhaps it must be interpreted differently;

- If we get this status, it's will only be related to authentication, if we passed one in the first place (?)
- OR there's no need to try getting a token if we don't have a username

@dmcgowan do you remember why "username" was checked for here? (and if it should only be relevant for a `405`, or treated the same for all of them?

Reply:
Looking at `FetchToken`, it looks like `UserName` is used for BasicAuth;

https://github.com/containerd/containerd/blob/8cc09e6a71de3f5b1b3739c1bed7df03ff7931c1/remotes/docker/auth/fetch.go#L186-L188


So this check looks a bit suspicious (if I have a `405` but don't have `to.Secret`, the `to.UserName` is not used; would all of this apply equally to other statuses?)

At least it looks like we need better documentation of this code.

Reply:
The `GET` request uses basic auth as your mentioned and there is no defined behavior for the endpoint with no username. Since this is for compatibility only, we don't want to try and add or define new behavior for that endpoint. I think you are right in that the right behavior would be to not fallback under any status when no username is defined. Documentation could make that clearer.